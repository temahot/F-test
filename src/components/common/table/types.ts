import { type UsePaginationReturn } from '~/hooks/usePagination';

export interface TableProps {
  data: Array<Record<string, any>> | [];
  headers: string[];
  pagination?: UsePaginationReturn['pagination'];
  itemsAmount?: number;
  isLoading: boolean;
}
