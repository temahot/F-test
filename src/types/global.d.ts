declare type Nullable<T> = T | null;

declare type ExtractEnums<T> = T extends ReadonlyArray<infer E> ? E : never;
