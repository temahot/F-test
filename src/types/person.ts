export type Person = {
  birth_year: string;
  eye_color: string;
  gender: string;
  hair_color: string;
  height: string;
  homeworld: string;
  mass: string;
  name: string;
  skin_color: string;
  url: string;
};

export type PersonPaginatedResponse = {
  count: number;
  next: string | null;
  previous: string | null;
  results: Person[];
};
