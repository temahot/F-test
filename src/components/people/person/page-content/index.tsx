import { useEffect } from 'react';

import ArrowBack from '@mui/icons-material/ArrowBack';
import { Box, Grid, styled } from '@mui/material';
import { useRouter } from 'next/router';
import { useSWRConfig } from 'swr';

import { Button } from '~/components/common/button';
import { TextInput } from '~/components/common/text-input';
import { PageTitle } from '~/components/page-title';
import usePersonValidation from '~/components/people/person/page-content/validation/usePersonValidation';
import { notify } from '~/helpers/common/notify';
import getNumberDecade from '~/helpers/getNumberDecade';
import usePerson from '~/hooks/people/person';
import { useProgress } from '~/providers/progress-provider';
import { ROUTES } from '~/routes';
import { type Person } from '~/types/person';

const ActionGroup = styled(Box)({
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
  justifyContent: 'flex-end',
  marginRight: 15,
});

export default function PersonPageContent({ personId }: { personId: number }) {
  const router = useRouter();
  const { mutate, cache } = useSWRConfig();
  const { setProgressIsShownHandler } = useProgress();
  const { person, personIsLoading } = usePerson({ id: personId });
  const { form, resetPersonForm } = usePersonValidation();
  const newPerson = form.watch();
  const firstPersonLoading = personIsLoading && person === undefined;

  const onSaveHandler = form.handleSubmit(async () => {
    setProgressIsShownHandler(true);
    if (person === undefined) return;

    const cachePage = getNumberDecade(personId) + 1;
    const revalidatePath = `${ROUTES.PEOPLE}?page=${cachePage}`;
    const peopleCache = cache.get(revalidatePath);
    if (peopleCache) {
      const foundPersonIndex = peopleCache.data.results.findIndex(
        (item: Person) => item.url === person.url
      );
      peopleCache.data.results[foundPersonIndex] = newPerson;
      notify('success', 'Saved');
    }

    await mutate(`${ROUTES.PEOPLE}/${personId}`, newPerson, {
      optimisticData: true,
      revalidate: false,
    });

    setProgressIsShownHandler(false);
    router.back();
  });

  useEffect(() => {
    if (person) {
      resetPersonForm(person);
      return;
    }
    void mutate(`${ROUTES.PEOPLE}/${personId}`);
  }, [mutate, person, personId, resetPersonForm]);

  return (
    <Box>
      <Button
        startIcon={<ArrowBack />}
        color={'info'}
        onClick={() => {
          void router.push(`/${ROUTES.PEOPLE}`);
        }}
      >
        Back
      </Button>

      <PageTitle title={'Change person'} />

      <Grid container>
        <Grid item sm={4}>
          <TextInput
            fullWidth
            form={form}
            name={'name'}
            isLoading={firstPersonLoading}
          />
          <Box display={'flex'} justifyContent={'space-between'}>
            <TextInput
              fullWidth
              form={form}
              name={'height'}
              type={'number'}
              isLoading={firstPersonLoading}
              isNumber
            />
            <TextInput
              fullWidth
              form={form}
              name={'mass'}
              type={'number'}
              isLoading={firstPersonLoading}
              isNumber
            />
          </Box>

          <TextInput
            fullWidth
            form={form}
            name={'gender'}
            isLoading={firstPersonLoading}
          />
        </Grid>

        <Grid item sm={8}>
          <TextInput
            fullWidth
            form={form}
            name={'eye_color'}
            isLoading={firstPersonLoading}
          />
          <TextInput
            fullWidth
            form={form}
            name={'skin_color'}
            isLoading={firstPersonLoading}
          />

          <TextInput
            fullWidth
            form={form}
            name={'hair_color'}
            isLoading={firstPersonLoading}
          />
        </Grid>
        <Grid container>
          <Grid sm item>
            <TextInput
              fullWidth
              form={form}
              name={'homeworld'}
              isLoading={firstPersonLoading}
            />
            <ActionGroup>
              <Button
                sx={{ mr: 1 }}
                variant={'outlined'}
                color="secondary"
                onClick={() => {
                  if (person) resetPersonForm(person);
                }}
              >
                reset
              </Button>
              <Button
                disabled={!router.isReady}
                color="primary"
                onClick={async () => {
                  await onSaveHandler();
                }}
                variant={'outlined'}
              >
                save
              </Button>
            </ActionGroup>
          </Grid>
        </Grid>
      </Grid>
    </Box>
  );
}
