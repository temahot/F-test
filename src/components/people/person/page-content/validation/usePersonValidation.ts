import { useCallback } from 'react';
import { useForm } from 'react-hook-form';

import { yupResolver } from '@hookform/resolvers/yup';
import * as y from 'yup';

import { type Person } from '~/types/person';

const numValidation = y
  .number()
  .positive('More then one')
  .required('Please add value')
  .transform((value) => (isNaN(value) || value === undefined ? null : value));

export default function usePersonValidation() {
  const validation = y.object().shape({
    name: y.string().required(),
    height: numValidation,
    mass: numValidation,
    eye_color: y.string().required(),
    gender: y.string().required(),
    hair_color: y.string().required(),
    homeworld: y.string().required(),
    skin_color: y.string().required(),
  });

  const defaultValues = {
    eye_color: '',
    gender: '',
    hair_color: '',
    height: '',
    homeworld: '',
    mass: '',
    name: '',
    skin_color: '',
  };

  const form = useForm<Person>({
    defaultValues,
    mode: 'all',
    resolver: yupResolver(validation),
  });

  const resetPersonForm = useCallback(
    (defaultValues: Person) => {
      form.reset(defaultValues, {
        keepDefaultValues: true,
      });
    },
    [form]
  );

  return {
    form,
    defaultValues,
    resetPersonForm,
  };
}
