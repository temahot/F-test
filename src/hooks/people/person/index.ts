import { useMemo } from 'react';

import { useRouter } from 'next/router';
import useSWR from 'swr';

import { notify } from '~/helpers/common/notify';
import useNetwork from '~/hooks/common/useNetwork';
import { ROUTES } from '~/routes';
import { type Person } from '~/types/person';

interface UsePersonReturn {
  personIsLoading: boolean;
  person: Person | undefined;
}

interface UsePersonProps {
  id: number | string;
}

export default function usePerson({ id }: UsePersonProps): UsePersonReturn {
  const { api } = useNetwork();
  const router = useRouter();
  const fetchPersonById = async () => {
    try {
      const response = await api.people.fetchPeopleById({
        id,
      });
      return response.data;
    } catch {
      notify('error', 'Person not found');
      await router.push(`/${ROUTES.PEOPLE}`);
      throw new Error('fetchPersonById failed');
    }
  };

  const { data, isLoading } = useSWR<Person>(
    `${ROUTES.PEOPLE}/${id}`,
    fetchPersonById,
    {
      revalidateOnFocus: false,
      revalidateOnMount: false,
    }
  );

  return useMemo(
    () => ({
      personIsLoading: isLoading,
      person: data,
    }),
    [data, isLoading]
  );
}
