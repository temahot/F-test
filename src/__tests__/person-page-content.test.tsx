import React from 'react';
import { useController } from 'react-hook-form';
import renderer, { act } from 'react-test-renderer';

import { render } from '@testing-library/react';
import { useRouter } from 'next/router';
import { useSWRConfig } from 'swr';

import allPeopleResponse from '~/__mocks__/all-people-response';
import PersonPageContent from '~/components/people/person/page-content';
import usePersonValidation from '~/components/people/person/page-content/validation/usePersonValidation';
import usePerson from '~/hooks/people/person';
import { type Person } from '~/types/person';

const mockData: { person: Person; testCb: () => void } = {
  person: allPeopleResponse.results[0],
  testCb: jest.fn(),
};

jest.mock('react-hook-form');
jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));
jest.mock(
  '~/components/people/person/page-content/validation/usePersonValidation'
);
jest.mock('~/hooks/people/person');
jest.mock('swr', () => ({
  mutate: jest.fn(),
  useSWRConfig: jest.fn(),
}));

describe('Person page component', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
    mockData.person = allPeopleResponse.results[0];
    mockData.testCb = jest.fn();
    (useController as jest.Mock).mockReturnValue({
      field: jest.fn(),
    });
    (useRouter as jest.Mock).mockReturnValue({ query: {} });
    (usePerson as jest.Mock).mockReturnValue({
      person: mockData.person,
      personIsLoading: false,
    });
    (usePersonValidation as jest.Mock).mockReturnValue({
      resetPersonForm: mockData.testCb,
      form: {
        watch: jest.fn().mockReturnValue(mockData.person),
        handleSubmit: jest.fn(),
        formState: {
          errors: [],
        },
      },
    });
    (useSWRConfig as jest.Mock).mockReturnValue({
      mutate: jest.fn(),
    });

    await act(() => {
      render(<PersonPageContent personId={1} />);
    });
  });

  it('Renders correctly', () => {
    const tree = renderer.create(<PersonPageContent personId={1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('resetFrom should call mutate if person exists', () => {
    expect(mockData.testCb).toBeCalledWith(mockData.person);
  });
});
