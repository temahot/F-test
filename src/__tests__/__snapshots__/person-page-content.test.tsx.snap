// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person page component Renders correctly 1`] = `
<div
  className="MuiBox-root css-0"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeMedium MuiButton-textSizeMedium css-1hny3k0-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ArrowBackIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
        />
      </svg>
    </span>
    Back
  </button>
  <h4
    className="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
  >
    Change person
  </h4>
  <div
    className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4 css-kgs8gf-MuiGrid-root"
    >
      <div
        className="MuiBox-root css-1vzneqq"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r0:"
            id=":r0:-label"
          >
            Name
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r0:"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <h6
          className="MuiTypography-root MuiTypography-subtitle2 css-lym8t0-MuiTypography-root"
        />
      </div>
      <div
        className="MuiBox-root css-gg4vpm"
      >
        <div
          className="MuiBox-root css-1vzneqq"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor=":r1:"
              id=":r1:-label"
            >
              Height
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id=":r1:"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="number"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    Height
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <h6
            className="MuiTypography-root MuiTypography-subtitle2 css-lym8t0-MuiTypography-root"
          />
        </div>
        <div
          className="MuiBox-root css-1vzneqq"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor=":r2:"
              id=":r2:-label"
            >
              Mass
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id=":r2:"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="number"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    Mass
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <h6
            className="MuiTypography-root MuiTypography-subtitle2 css-lym8t0-MuiTypography-root"
          />
        </div>
      </div>
      <div
        className="MuiBox-root css-1vzneqq"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r3:"
            id=":r3:-label"
          >
            Gender
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r3:"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Gender
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <h6
          className="MuiTypography-root MuiTypography-subtitle2 css-lym8t0-MuiTypography-root"
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-8 css-3b1rzb-MuiGrid-root"
    >
      <div
        className="MuiBox-root css-1vzneqq"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r4:"
            id=":r4:-label"
          >
            Eye color
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r4:"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Eye color
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <h6
          className="MuiTypography-root MuiTypography-subtitle2 css-lym8t0-MuiTypography-root"
        />
      </div>
      <div
        className="MuiBox-root css-1vzneqq"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r5:"
            id=":r5:-label"
          >
            Skin color
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r5:"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Skin color
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <h6
          className="MuiTypography-root MuiTypography-subtitle2 css-lym8t0-MuiTypography-root"
        />
      </div>
      <div
        className="MuiBox-root css-1vzneqq"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r6:"
            id=":r6:-label"
          >
            Hair color
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r6:"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Hair color
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <h6
          className="MuiTypography-root MuiTypography-subtitle2 css-lym8t0-MuiTypography-root"
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-true css-1vknvy4-MuiGrid-root"
      >
        <div
          className="MuiBox-root css-1vzneqq"
        >
          <div
            className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor=":r7:"
              id=":r7:-label"
            >
              Homeworld
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id=":r7:"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    Homeworld
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <h6
            className="MuiTypography-root MuiTypography-subtitle2 css-lym8t0-MuiTypography-root"
          />
        </div>
        <div
          className="MuiBox-root css-1586a6o"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-y58v9j-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            reset
          </button>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            save
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
